nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecordTest.java

```diff
@@ -0,0 +1,850 @@
+
+        if (success > 0) {
+            runner.getFlowFilesForRelationship(PutElasticsearchRecord.REL_SUCCESS).forEach(ff -> {
+                ff.assertAttributeEquals("elasticsearch.put.success.count", "0");
+                ff.assertAttributeEquals("elasticsearch.put.error.count", "0");});
+
+            assertEquals(success,
+                    runner.getProvenanceEvents().stream().filter(
+                            e -> ProvenanceEventType.SEND.equals(e.getEventType()) && "1 Elasticsearch _bulk operation batch(es) [0 error(s), 0 success(es)]".equals(e.getDetails()))
+                            .count());
+        }
+    }
+
+    @Test
+    public void simpleTest() {
+        clientService.setEvalParametersConsumer((Map<String, String> params) -> assertTrue(params.isEmpty()));
+        basicTest(0, 0, 1);
+    }
+
+    @Test
+    public void simpleTestCoercedDefaultTimestamp() {
+        Consumer<List<IndexOperationRequest>> consumer = (List<IndexOperationRequest> items) ->
+                assertEquals(2L, items.stream().filter(item -> Long.valueOf(100).equals(item.getFields().get("@timestamp"))).count());
+
+        runner.setProperty(PutElasticsearchRecord.AT_TIMESTAMP, "100");
+        basicTest(0, 0, 1, consumer);
+    }
+
+    @Test
+    public void simpleTestWithRequestParametersAndBulkHeaders() {
+        runner.setProperty("refresh", "true");
+        runner.setProperty(AbstractPutElasticsearch.BULK_HEADER_PREFIX + "routing", "/routing");
+        runner.setProperty(AbstractPutElasticsearch.BULK_HEADER_PREFIX + "version", "${version}");
+        runner.setProperty(AbstractPutElasticsearch.BULK_HEADER_PREFIX + "empty", "${empty}");
+        runner.setProperty("slices", "${slices}");
+        runner.setProperty("another", "${blank}");
+        runner.setVariable("slices", "auto");
+        runner.setVariable("blank", " ");
+        runner.setVariable("version", "/version");
+        runner.assertValid();
+
+        clientService.setEvalParametersConsumer( (Map<String, String> params) -> {
+            assertEquals(2, params.size());
+            assertEquals("true", params.get("refresh"));
+            assertEquals("auto", params.get("slices"));
+        });
+
+        Consumer<List<IndexOperationRequest>> consumer = (List<IndexOperationRequest> items) -> {
+            long headerFieldsCount = items.stream().filter(item -> !item.getHeaderFields().isEmpty()).count();
+            long routingCount = items.stream().filter(item -> "1".equals(item.getHeaderFields().get("routing"))).count();
+            long versionCount = items.stream().filter(item -> "external".equals(item.getHeaderFields().get("version"))).count();
+            assertEquals(2, headerFieldsCount);
+            assertEquals(1, routingCount);
+            assertEquals(1, versionCount);
+        };
+
+        basicTest(0, 0, 1, consumer);
+    }
+
+    @Test
+    public void simpleTestWithRequestParametersAndBulkHeadersFlowFileEL() {
```

Comment:
Pretty much the same as `simpleTestWithRequestParametersAndBulkHeaders`, consider creating a common `testWithRequestParametersAndBulkHeaders`, then have these test cases set test specific properties, etc. (think the main difference is that one passes a `Map` of EL values)